File: src/cli/journal.rs
use crate::journal::{JournalEntry, JournalFilter, TradeJournal};
use anyhow::Result;

pub fn run(
    symbol: Option<String>,
    event_type: Option<String>,
    limit: usize,
    paper_only: bool,
) -> Result<()> {
    let journal = TradeJournal::new(None);
    let filter = JournalFilter {
        symbol,
        event_type,
        limit: Some(limit),
        is_paper: if paper_only { Some(true) } else { None },
        ..Default::default()
    };

    let entries = journal.query(&filter)?;
    println!("\n{}", "=".repeat(125));
    println!("📖 BHyper Detailed Trade Execution Journal (Chronological Ledger)");
    println!("{}", "=".repeat(125));
    println!(
        "{:<20} {:<10} {:<8} {:<10} {:<45} {:<18}",
        "Timestamp (UTC)", "Event", "Symbol", "Mode", "Details / Execution", "PnL / Value"
    );
    println!("{}", "-".repeat(125));

    if entries.is_empty() {
        println!("  (No journal entries matching query filter found)");
    } else {
        for e in entries {
            let mode_tag = if e.is_paper() {
                "🧪 PAPER"
            } else {
                "⚡ LIVE"
            };
            let t_str = e.timestamp().format("%Y-%m-%d %H:%M:%S").to_string();

            match e {
                JournalEntry::Intent(i) => {
                    println!(
                        "{:<20} {:<10} {:<8} {:<10} {:<45} ${:<17.2}",
                        t_str,
                        "INTENT",
                        i.symbol,
                        mode_tag,
                        format!(
                            "Spread: {:.1}% | HL:{} BN:{}",
                            i.net_spread_apr_pct, i.hyperliquid_side, i.binance_side
                        ),
                        i.target_notional_usd
                    );
                }
                JournalEntry::OpenFill(o) => {
                    println!(
                        "{:<20} {:<10} {:<8} {:<10} {:<45} ${:<17.2}",
                        t_str,
                        "OPEN_FILL",
                        o.symbol,
                        mode_tag,
                        format!(
                            "HL: ${:.4} | BN: ${:.4} (Fee: ${:.4})",
                            o.hyperliquid_price, o.binance_price, o.total_open_fees_usd
                        ),
                        o.total_notional_usd
                    );
                }
                JournalEntry::Funding(f) => {
                    println!(
                        "{:<20} {:<10} {:<8} {:<10} {:<45} ${:<17.4}",
                        t_str,
                        "FUNDING",
                        f.symbol,
                        mode_tag,
                        format!(
                            "{}: {:.2} bps | Cum: ${:.4}",
                            f.exchange, f.rate_bps, f.cumulative_funding_usd
                        ),
                        f.funding_payment_usd
                    );
                }
                JournalEntry::RiskAlert(r) => {
                    println!(
                        "{:<20} {:<10} {:<8} {:<10} {:<45} {:<18}",
                        t_str,
                        "RISK",
                        r.symbol,
                        mode_tag,
                        format!("{}: {}", r.event_type, r.details),
                        r.action_taken
                    );
                }
                JournalEntry::CloseFill(c) => {
                    println!(
                        "{:<20} {:<10} {:<8} {:<10} {:<45} ${:<17.4}",
                        t_str,
                        "CLOSE_FILL",
                        c.symbol,
                        mode_tag,
                        format!(
                            "Held: {:.1}h | Funding: ${:.4} | Fees: ${:.4}",
                            c.holding_duration_secs as f64 / 3600.0,
                            c.gross_funding_earned_usd,
                            c.total_roundtrip_fees_usd
                        ),
                        c.net_realized_pnl_usd
                    );
                }
            }
        }
    }
    println!("{}\n", "=".repeat(125));
    Ok(())
}

File: static/js/components/journal.js
/**
 * BHyper Terminal - Holographic Journal Component
 * Pure English Institutional Layout
 */
import { apiFetch } from '../api.js';
import { formatCurrency, formatPrice, formatTimeOnlyUtc8 } from '../utils/format.js';

export async function fetchAndRenderJournal(filterType = '') {
  const tbody = document.getElementById('journal-table-body');
  if (!tbody) return;

  try {
    const res = await apiFetch(`/api/journal?event_type=${encodeURIComponent(filterType)}&limit=50`);
    if (!res.entries || res.entries.length === 0) {
      tbody.innerHTML = `
        <tr>
          <td colspan="7" class="text-center py-12 text-[var(--text-muted)]">
            <div class="flex flex-col items-center justify-center space-y-2">
              <i data-lucide="file-text" class="w-6 h-6 text-slate-500 opacity-60"></i>
              <span class="text-xs">No ledger entries recorded yet.</span>
            </div>
          </td>
        </tr>
      `;
      return;
    }

    tbody.innerHTML = res.entries.map(e => {
      const typeName = e.type;
      const d = e.data;
      const time = formatTimeOnlyUtc8(d.timestamp);

      let typeBadge = `<span class="px-2 py-0.2 rounded text-[10px] font-medium bg-slate-200 dark:bg-slate-800 text-[var(--text-primary)]">${typeName}</span>`;
      let detail = '';
      let notional = '-';
      let pnlOrFunding = '-';
      let fees = '-';

      if (typeName === 'Funding') {
        typeBadge = `<span class="px-2 py-0.2 rounded text-[10px] font-medium bg-cyan-500/10 text-cyan-500 border border-cyan-500/25">FUNDING</span>`;
        detail = `${d.exchange || ''} ${d.side || ''} (Rate: ${(d.rate_bps || 0).toFixed(2)} bps)`;
        notional = formatCurrency(d.notional_usd || 0);
        const pay = d.funding_payment_usd || 0;
        pnlOrFunding = `<span class="${pay >= 0 ? 'text-cyan-500' : 'text-rose-500'} font-semibold">${pay >= 0 ? '+' : ''}${formatCurrency(pay, 4)}</span>`;
        fees = '$0.0000';
      } else if (typeName === 'OpenFill') {
        typeBadge = `<span class="px-2 py-0.2 rounded text-[10px] font-medium bg-emerald-500/10 text-emerald-500 border border-emerald-500/25">OPEN</span>`;
        detail = `HL: ${d.hyperliquid_side} ${formatPrice(d.hyperliquid_price)} / BN: ${d.binance_side} ${formatPrice(d.binance_price)}`;
        notional = formatCurrency(d.total_notional_usd || d.notional_usd || 0);
        pnlOrFunding = `<span class="text-slate-400">Fully Hedged</span>`;
        fees = formatCurrency(d.total_open_fees_usd || 0, 4);
      } else if (typeName === 'CloseFill') {
        typeBadge = `<span class="px-2 py-0.2 rounded text-[10px] font-medium bg-rose-500/10 text-rose-500 border border-rose-500/25">CLOSE</span>`;
        detail = `HL: ${formatPrice(d.hyperliquid_exit_price)} / BN: ${formatPrice(d.binance_exit_price)} (Basis PnL: ${formatCurrency(d.gross_basis_pnl_usd || 0, 4)})`;
        const pnlVal = d.net_realized_pnl_usd || 0;
        pnlOrFunding = `<span class="${pnlVal >= 0 ? 'text-emerald-500' : 'text-rose-500'} font-semibold">${pnlVal >= 0 ? '+' : ''}${formatCurrency(pnlVal, 4)}</span>`;
        fees = formatCurrency(d.total_roundtrip_fees_usd || 0, 4);
      } else if (typeName === 'Intent') {
        typeBadge = `<span class="px-2 py-0.2 rounded text-[10px] font-medium bg-purple-500/10 text-purple-500 border border-purple-500/25">INTENT</span>`;
        detail = `${d.reason || 'Arbitrage Trigger Signal'}`;
        notional = formatCurrency(d.target_notional_usd || 0);
        pnlOrFunding = `<span class="text-emerald-500 font-medium">+${(d.net_spread_apr_pct || 0).toFixed(2)}% APR</span>`;
        fees = '-';
      } else if (typeName === 'RiskAlert') {
        typeBadge = `<span class="px-2 py-0.2 rounded text-[10px] font-medium bg-amber-500/10 text-amber-500 border border-amber-500/25">RISK</span>`;
        detail = `${d.event_type || ''}: ${d.details || ''} (${d.action_taken || ''})`;
        pnlOrFunding = `<span class="text-amber-500 font-medium">Risk Guard</span>`;
      }

      return `
        <tr class="hover:bg-[var(--table-hover)] transition">
          <td class="py-2.5 px-3.5 text-[var(--text-muted)]">${time}</td>
          <td class="py-2.5 px-2">${typeBadge}</td>
          <td class="py-2.5 px-2 font-semibold text-[var(--text-primary)]">${d.symbol}</td>
          <td class="py-2.5 px-3 text-[var(--text-secondary)]">${detail}</td>
          <td class="py-2.5 px-2 text-right text-[var(--text-muted)]">${notional}</td>
          <td class="py-2.5 px-2 text-right">${pnlOrFunding}</td>
          <td class="py-2.5 px-3 text-right ${fees.startsWith('$') && fees !== '$0.0000' ? 'text-rose-500' : 'text-[var(--text-muted)]'}">${fees.startsWith('$') && fees !== '$0.0000' ? '-' + fees : fees}</td>
        </tr>
      `;
    }).join('');

    if (window.lucide) {
      window.lucide.createIcons();
    }
  } catch (e) {
    console.warn('Journal error:', e);
  }
}
